### DISCUSSION TITLE
HTTPS redirect middleware causing infinite redirect loop on Heroku

### DISCUSSION THREAD

**Discussion by samhastings** [vercel/next.js#51970] [Help] [answered]
### Summary

I've been successfully using this `middleware.ts` code for a while now on several Next.js projects to redirect all http requests to https on Heroku. I've just deployed a brand new project to Heroku and it's now causing a "this page is not redirecting properly" error as https requests are also being redirected causing an infinite loop.

```typescript
import { NextFetchEvent, NextRequest, NextResponse } from 'next/server';

export function middleware(req: NextRequest, ev: NextFetchEvent) {
    if (process.env.NODE_ENV === 'production' && req.headers.get('x-forwarded-proto') !== 'https') {
        return NextResponse.redirect(`https://${req.headers.get('host')}${req.nextUrl.pathname}`, 301);
    }

    return NextResponse.next();
}
```

I'm guessing something's changed at Heroku's end that might be making this happen all of a sudden, but does anyone know of a workaround? Happy to use a custom server if absolutely necessary but prefer to avoid the added complexity this brings!

**Comment by alphabeat** [comment, marked as answer]
Hello @samhastings,

We have had the exact same issue since we've upgraded to nextjs v13.4.7. After some tests on a Heroku staging environment, we found out that `req.headers.get('x-forwarded-proto')` now returns "http,http" or "https,http" rather than "http" or "https" (which should be the expected value according to the [MDN reference](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-Proto)).

It looks like this is a nextjs update and not a Heroku problem. To fix your issue, I suggest this change in your code (it fixed the problem on our end):
```js
if (process.env.NODE_ENV === 'production' && !req.headers.get('x-forwarded-proto').includes('https')) {
...
}
```

**Comment by samhastings** [comment]
Funnily enough I managed to get to the bottom of this issue a couple of days ago and planned to post the solution here but you beat me to it! I'm surprised it's a Next.js issue – assumed something had changed at Heroku’s end – but glad to have it sorted all the same without having to resort to a custom server. Thanks for creating an issue to report this, I'll mark this as solved and hopefully this thread will prove useful to anyone else who's run into the same problem.
